File: cpp/sorting/bubble_sort/bubble_sort.cpp
//
// Created by mullins on 3/22/19.
//
// C++ Bubble Sort
#include <stdio.h>

void Swap(int *xp, int *yp) {
    int temp = *xp;
    *xp = *yp;
    *yp = temp;
}

void Swap(char *xp, char *yp) {
    char temp = *xp;
    *xp = *yp;
    *yp = temp;
}

void BubbleSort(int arr[], int len) {
    int i, j;

    for (i = 0; i < len - 1; i++) {
        // Last i elements are already in place
        for (j = 0; j < len - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                Swap(&arr[j], &arr[j + 1]);
            }
        }
    }
}

void BubbleSort(char arr[], int len) {
    int i, j;

    for (i = 0; i < len - 1; i++) {
        // Last i elements are already in place
        for (j = 0; j < len - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                Swap(&arr[j], &arr[j + 1]);
            }
        }
    }
}

/* Functions to print array */
void PrintArray(int arr[], int len) {
    int i;
    for (i = 0; i < len; i++)
        printf("%d, ", arr[i]);
    printf("\n");
}

void PrintArray(char arr[], int len) {
    int i;
    for (i = 0; i < len; i++)
        printf("%c, ", arr[i]);
    printf("\n");
}

int main() {
    int intArr[] = {4, 6, 3, 2, 6, 8, 1, 5, 9, 5, 2, 3, 8, 1, 2, 5, 0, 9};
    int intLen = sizeof(intArr) / sizeof(intArr[0]);
    printf("Unsorted Array: \n");
    PrintArray(intArr, intLen);
    BubbleSort(intArr, intLen);
    printf("Sorted Array: \n");
    PrintArray(intArr, intLen);

    printf("\n");

    char charArr[] = {'a', 'n', 'e', 't', 'u', 'h', 'e', 's', 'x', 'h', 'i', 'o', 'j', 'g', 'f', 'd', 'e', 'g', 'z',
                      'a', 'e', 'w', 'f', 'h', 'k', 'h', 'f', 'l', 'a', 'q', 'w', 'r', 'y', 'h', 'f'};
    int charLen = sizeof(charArr) / sizeof(charArr[0]);
    printf("Unsorted Array: \n");
    PrintArray(charArr, charLen);
    BubbleSort(charArr, charLen);
    printf("Sorted Array: \n");
    PrintArray(charArr, charLen);
    return 0;
}

File: kotlin/sorting/bubble_sort/bubbleSort.kt
fun int_bubble_sort(arr: Array<Int>) {
    var len = arr.size
    var isSwapped = true
    while (isSwapped) {
        isSwapped = false
        for (i in 0 until len - 1) {
            if (arr[i] > arr[i+1]) {
                val temp = arr[i]
                arr[i] = arr[i+1]
                arr[i + 1] = temp

                isSwapped = true
            }
        }
    }
}

fun char_bubble_sort(arr: Array<Char>) {
    var len = arr.size
    var isSwapped = true
    while (isSwapped) {
        isSwapped = false
        for (i in 0 until len - 1) {
            if (arr[i] > arr[i+1]) {
                val temp = arr[i]
                arr[i] = arr[i+1]
                arr[i + 1] = temp

                isSwapped = true
            }
        }
    }
}

fun main(arg: Array<String>) {
    val intArr = arrayOf(4, 6, 3, 2, 6, 8, 1, 5, 9, 5, 2, 3, 8, 1, 2, 5, 0, 9)

    val charArr = arrayOf('a', 'n', 'e', 't', 'u', 'h', 'e', 's', 'x', 'h', 'i', 'o', 'j', 'g', 'f', 'd', 'e', 'g', 'z', 'a',
            'e', 'w', 'f', 'h', 'k', 'h', 'f', 'l', 'a', 'q', 'w', 'r', 'y', 'h', 'f')

    println("Unsorted Int Array: ")
    for (i in 0 until intArr.size)
        print("${intArr[i]},  ")
    println("\n")
    int_bubble_sort(intArr)
    println("Sorted Int Array: ")
    for (i in 0 until intArr.size)
        print("${intArr[i]},  ")

    println("\n\n")

    println("Unsorted Char Array: ")
    for (i in 0 until charArr.size)
        print("${charArr[i]},  ")
    println("\n")
    char_bubble_sort(charArr)
    println("Sorted Char Array: ")
    for (i in 0 until charArr.size)
        print("${charArr[i]},  ")
}